File: mmv1/provider/tflint.rb
module Provider
  class TFLint < Provider::Terraform
    @@rule_names = []
    @@resource_url = {}

    def generate_resource(pwd, data, generate_code, generate_docs)
      @terraform_name = data.object.legacy_name || "google_#{full_resource_name(data)}"

      begin
        @@resource_url[@terraform_name] = URI.parse(data.product.base_url).host
      rescue URI::InvalidURIError => exn
        Google::LOGGER.warn "Cannot parse Api::Product#base_url: #{exn}"
      end

      data.object.all_user_properties.each do |prop|
        next if prop.output
        next unless prop.is_a?(Api::Type::Enum) || prop.validation&.regex

        @prop = prop
        @rule_name = "#{@terraform_name}_invalid_#{prop.name.underscore}"

        data.generate(pwd, '/templates/tflint/rule.go.erb', "#{@rule_name}.go", self)

        @@rule_names << @rule_name
      end
    end

    def generate_resource_tests(pwd, data) end

    def generate_resource_sweepers(pwd, data) end

    def generate_iam_policy(pwd, data, generate_code, generate_docs) end

    def generate_operation(pwd, output_folder, _types) end

    def compile_common_files(output_folder, products, common_compile_file)
      @rule_names = @@rule_names.sort
      @resource_url = @@resource_url

      Google::LOGGER.info 'Compiling common files.'
      file_template = ProviderFileTemplate.new(
        output_folder,
        @target_version_name,
        build_env,
        products
      )
      compile_file_list(output_folder, [['provider.go', 'templates/tflint/provider.go.erb']], file_template)
      compile_file_list(output_folder, [['api_definition.go', 'templates/tflint/api_definition.go.erb']], file_template)
    end

    def copy_common_files(output_folder, generate_code, generate_docs)
      Google::LOGGER.info 'Copying common files.'
      copy_file_list(output_folder, [
                       ['verify/validation.go',
                        'third_party/terraform/verify/validation.go'],
                    ])
    end
  end
end

File: mmv1/provider/tflint.go
package provider

import (
	"fmt"
	"log"
	"net/url"
	"os"
	"path"
	"sort"
	"time"

	"github.com/GoogleCloudPlatform/magic-modules/mmv1/api"
	"github.com/GoogleCloudPlatform/magic-modules/mmv1/google"
)

type TFLint struct {
	Terraform
}

func NewTFLint(product *api.Product, versionName string, startTime time.Time) TFLint {
	return TFLint{
		Terraform: NewTerraform(product, versionName, startTime),
	}
}

func (t TFLint) Generate(outputFolder, productPath, resourceToGenerate string, generateCode, generateDocs bool) {
	if err := os.MkdirAll(outputFolder, os.ModePerm); err != nil {
		log.Println(fmt.Errorf("error creating output directory %v: %v", outputFolder, err))
	}

	t.GenerateObjects(outputFolder, resourceToGenerate, generateCode, generateDocs)
}

func (t *TFLint) GenerateObjects(outputFolder, resourceToGenerate string, generateCode, generateDocs bool) {
	for _, object := range t.Product.Objects {
		object.ExcludeIfNotInVersion(&t.Version)

		if resourceToGenerate != "" && object.Name != resourceToGenerate {
			log.Printf("Excluding %s per user request", object.Name)
			continue
		}

		t.GenerateObject(*object, outputFolder, t.TargetVersionName, generateCode, generateDocs)
	}
}

func (t *TFLint) GenerateObject(object api.Resource, outputFolder, productPath string, generateCode, generateDocs bool) {
	templateData := NewTemplateData(outputFolder, t.TargetVersionName)

	if !object.IsExcluded() {
		log.Printf("Generating %s rules", object.Name)
		t.GenerateResource(object, *templateData, outputFolder, generateCode, generateDocs)
	}
}

func (t *TFLint) GenerateResource(object api.Resource, templateData TemplateData, outputFolder string, generateCode, generateDocs bool) {
	for _, prop := range object.AllUserProperties() {
		if !t.isTargetProp(prop) {
			continue
		}

		ruleName := t.generateRuleName(object, prop)
		targetFilePath := path.Join(outputFolder, fmt.Sprintf("%s.go", ruleName))

		input := struct {
			RuleName      string
			Rule_name     string
			ResourceType  string
			AttributeName string
			Prop          *api.Type
		}{
			RuleName:      google.Camelize(ruleName, "upper"),
			Rule_name:     ruleName,
			ResourceType:  object.TerraformName(),
			AttributeName: google.Underscore(prop.Name),
			Prop:          prop,
		}
		templateData.GenerateFile(targetFilePath, "templates/tflint/rule.go.tmpl", input, true, "templates/tflint/rule.go.tmpl")
	}
}

func (t *TFLint) isTargetProp(prop *api.Type) bool {
	if prop.Output {
		return false
	}
	if prop.IsA("Enum") {
		return true
	}
	if (prop.IsA("String") || prop.IsA("Integer")) && prop.Validation.Regex != "" {
		return true
	}
	return false
}

func (t *TFLint) generateRuleName(object api.Resource, prop *api.Type) string {
	return fmt.Sprintf("%s_invalid_%s", object.TerraformName(), google.Underscore(prop.Name))
}

func (t TFLint) CopyCommonFiles(outputFolder string, generateCode, generateDocs bool) {
	log.Printf("Copying common files for %s", ProviderName(t))

	files := map[string]string{
		"verify/validation.go": "third_party/terraform/verify/validation.go",
	}
	t.CopyFileList(outputFolder, files, generateCode)
}

func (t TFLint) CompileCommonFiles(outputFolder string, products []*api.Product, overridePath string) {
	templateData := NewTemplateData(outputFolder, t.TargetVersionName)

	type resourceURL struct {
		Name string
		URL  string
	}
	input := struct {
		RuleNames    []string
		ResourceURLs []resourceURL
	}{
		RuleNames:    []string{},
		ResourceURLs: []resourceURL{},
	}

	for _, product := range products {
		var productURL string
		u, err := url.Parse(product.BaseUrl)
		if err != nil {
			log.Println(fmt.Errorf("cannot parse product.BaseUrl: %v", err))
		}
		if u != nil {
			productURL = u.Host
		}

		for _, object := range product.Objects {
			object.ExcludeIfNotInVersion(&t.Version)
			if object.IsExcluded() {
				continue
			}

			if productURL != "" {
				input.ResourceURLs = append(input.ResourceURLs, resourceURL{Name: object.TerraformName(), URL: productURL})
			}

			for _, prop := range object.AllUserProperties() {
				if !t.isTargetProp(prop) {
					continue
				}

				ruleName := t.generateRuleName(*object, prop)
				input.RuleNames = append(input.RuleNames, google.Camelize(ruleName, "upper"))
			}
		}
	}
	sort.Strings(input.RuleNames)
	sort.Slice(input.ResourceURLs, func(i, j int) bool { return input.ResourceURLs[i].Name < input.ResourceURLs[j].Name })

	templateData.GenerateFile(
		path.Join(outputFolder, "provider.go"),
		"templates/tflint/provider.go.tmpl",
		input,
		true,
		"templates/tflint/provider.go.tmpl",
	)
	templateData.GenerateFile(
		path.Join(outputFolder, "api_definition.go"),
		"templates/tflint/api_definition.go.tmpl",
		input,
		true,
		"templates/tflint/api_definition.go.tmpl",
	)
}
